casper/src/test/scala/coop/rchain/casper/util/CasperUtilTest.scala

```diff
@@ -1,67 +1,60 @@
+      for {
+        genesis <- createBlock[StateWithChain](Seq())
+        b2      <- createBlock[StateWithChain](Seq(genesis.blockHash))
+        b3      <- createBlock[StateWithChain](Seq(b2.blockHash))
+      } yield b3
+
+    val initState =
+      Chain(HashMap.empty[Int, BlockMessage], HashMap.empty[ByteString, BlockMessage], 0)
+    val (chain: Chain, lastBlock: BlockMessage) = createChain.run(initState)
+
+    val genesis = chain.idToBlocks(1)
+    val b2      = chain.idToBlocks(2)
+    val b3      = chain.idToBlocks(3)
+    isInMainChain(chain.hashToBlocks, genesis, b3) should be(true)
+    isInMainChain(chain.hashToBlocks, b2, b3) should be(true)
+    isInMainChain(chain.hashToBlocks, b3, b2) should be(false)
+    isInMainChain(chain.hashToBlocks, b3, genesis) should be(false)
   }
 
   "isInMainChain" should "classify diamond DAGs appropriately" in {
-    val chain = mutable.HashMap[ByteString, BlockMessage]()
-
-    val genesis: BlockMessage = createBlock(Seq())
-    chain += (genesis.blockHash -> genesis)
-
-    val blockMessage2: BlockMessage = createBlock(Seq(genesis.blockHash))
-    chain += (blockMessage2.blockHash -> blockMessage2)
-
-    val blockMessage3: BlockMessage = createBlock(Seq(genesis.blockHash))
-    chain += (blockMessage3.blockHash -> blockMessage3)
-
-    val blockMessage4: BlockMessage =
-      createBlock(Seq(blockMessage2.blockHash, blockMessage3.blockHash))
-    chain += (blockMessage4.blockHash -> blockMessage4)
-
-    isInMainChain(chain, genesis, blockMessage2) should be(true)
-    isInMainChain(chain, genesis, blockMessage3) should be(true)
-    isInMainChain(chain, genesis, blockMessage4) should be(true)
-    isInMainChain(chain, blockMessage2, blockMessage4) should be(true)
-    isInMainChain(chain, blockMessage3, blockMessage4) should be(false)
-  }
-
-  private def createBlock(parentsHashList: Seq[ByteString]) = {
-    val resourceId     = resourceCounter.next()
-    val uniqueResource = Resource(ProduceResource(Produce(resourceId)))
-    val postState      = RChainState().withResources(Seq(uniqueResource))
-    val postStateHash  = Sha256.hash(postState.toByteArray)
-    val header = Header()
-      .withPostStateHash(ByteString.copyFrom(postStateHash))
-      .withParentsHashList(parentsHashList)
-    val blockHash = Sha256.hash(header.toByteArray)
-    val body      = Body().withPostState(postState)
-    BlockMessage(ByteString.copyFrom(blockHash), Some(header), Some(body))
+    val createChain =
+      for {
+        genesis <- createBlock[StateWithChain](Seq())
+        b2      <- createBlock[StateWithChain](Seq(genesis.blockHash))
+        b3      <- createBlock[StateWithChain](Seq(genesis.blockHash))
+        b4      <- createBlock[StateWithChain](Seq(b2.blockHash, b3.blockHash))
+      } yield b4
+
+    val initState =
+      Chain(HashMap.empty[Int, BlockMessage], HashMap.empty[ByteString, BlockMessage], 0)
+    val (chain: Chain, lastBlock: BlockMessage) = createChain.run(initState)
```

Comment:
Same here, maybe `runS` will suffice?

Reply:
Oh, sorry. I was sure it's gonna work. Sorry for misleading you. I guess `runS` comes from the `cats` while this is `scalaz`.